devito/operations/interpolators.py

```diff
@@ -388,3 +389,520 @@ def callback():
+                               evaluate=False)
+                condition = sympy.And(lb, ub, evaluate=False)
+                mapper[d] = ConditionalDimension(p.name, self.sfunction._sparse_dim,
+                                                 condition=condition, indirect=True)
+
+            # Track Indexed substitutions
+            idx_subs.append(mapper)
+
+        # Temporaries for the position
+        temps = [Eq(v, k, implicit_dims=self.sfunction.dimensions)
+                 for k, v in self.sfunction._position_map.items()]
+
+        temps.extend([Eq(v, k.subs(self.sfunction._position_map),
+                      implicit_dims=self.sfunction.dimensions)
+                     for k, v in self.sfunction._relative_position_map.items()])
+
+        # Temporaries for the indirection dimensions
+        temps.extend([Eq(v, k.subs(self.sfunction._position_map),
+                         implicit_dims=self.sfunction.dimensions)
+                      for k, v in points.items()])
+
+        return idx_subs, temps
+
+    def interpolate(self, expr, offset=0, increment=False, self_subs={}):
+        """
+        Generate equations interpolating an arbitrary expression into ``self``.
+        Parameters
+        ----------
+        expr : expr-like
+            Input expression to interpolate.
+        offset : int, optional
+            Additional offset from the boundary.
+        increment: bool, optional
+            If True, generate increments (Inc) rather than assignments (Eq).
+        """
+        def callback():
+
+            # Derivatives must be evaluated before the introduction of indirect accesses
+            try:
+                _expr = expr.evaluate
+            except AttributeError:
+                # E.g., a generic SymPy expression or a number
+                _expr = expr
+
+            variables = list(retrieve_function_carriers(_expr))
+
+            # Need to get origin of the field in case it is staggered
+            # TODO: handle each variable staggereing spearately
+            field_offset = variables[0].origin
+
+            # List of indirection indices for all adjacent grid points
+            idx_subs, temps = self._interpolation_indices(variables, offset,
+                                                          field_offset=field_offset)
+
+            rhs = Symbol(name='sum', dtype=self.sfunction.dtype)
+            summands = [Eq(rhs, 0., implicit_dims=self.sfunction.dimensions)]
+
+            dim_pos = self.sfunction.grid.dimensions
+
+            # Checks if the data is 3D or 2D
+            if len(dim_pos) == 3:
+                eqs = [np.sum(v) for v in self._tricubic_equations(_expr,
+                                                                   idx_subs,
+                                                                   dim_pos=dim_pos)]
+            else:
+                eqs = [np.sum(v) for v in self._bicubic_equations(_expr,
```

Comment:
why doesn't `..._equations` directly return the sum?
Also why `np.sum`? Isn't it still partially symbolic?

Reply:
Not returning directly the sum, the return of `..._equations` can be used for both injection and interpolation. In interpolation it sum the values and in injection just assign the values for its respective grid point.

About the use of numpy: Yeah, it is still a symbolic. I wanted to generate a symbolic equation that sum all the individual values, that's why i used `np.sum`. Is there some problem in using it?

Reply:
Does this actually work? Numpy summing over sympy symbols?

Reply:
Yes, it works summing over sympy symbols. But I will take a look at Devito.Add(*v) to use it